added EditAction::SelectNone

sysint64 opened the pull request:
this PR adds a new edit action SelectNone as the opposite of SelectAll.

sysint64 commented:
@cmyr okay, I got it.

cmyr commented:
@sysint64 ping on this?

sysint64 commented:
@cmyr I'm still here, just a lot of work, I'll work on my PR's on the following weekends.

sysint64 commented:
@cmyr does `Movement::LeftOfLine` do the same thing as should `BeginningOfDocument`? at least the current implementation seems to me what I have tried to do.

cmyr commented:
It may be the same now, but it shouldn't be the same when we are handling multi-line text correctly; if it is that's likely a bug in the current implementation.

sysint64 commented:
@cmyr Sorry for the delay, I have implemented two methods - `line_beginning` and `line_ending` for `EditableText`, and added `Movement::BeginningOfDocument` and `Movement::EndingOfDocument`.

cmyr commented on druid/src/text/movement.rs:
```diff
@@ -31,6 +31,10 @@ pub enum Movement {
+    /// Move to the beginning of the document
+    BeginningOfDocument,
+    /// Move to the end of the document
+    EndingOfDocument,
```
I would call these `StartOfDocument` and `EndOfDocument`.

cmyr commented on druid/src/text/editable_text.rs:
```diff
@@ -58,6 +58,12 @@ pub trait EditableText: Sized {
+
+    /// Get the line ending offset from the given offset, if it exists.
+    fn line_ending(&self, offset: usize) -> Option<usize>;
+
```
Let's call these `left_end_of_line` and `right_end_of_line`. 'beginning' and 'end' are ambiguous, since in right-to-left languages the beginning of the line is on the right. :)

cmyr commented on druid/src/text/editable_text.rs:
```diff
@@ -156,6 +162,32 @@ impl EditableText for String {
+                return Some(offset);
+            }
+            offset -= grapheme.len();
+        }
```
We don't need to be fancy with graphemes here; we're only looking for the 0x80 (`\n`) byte, and we can must iterate the bytes directly, which is much cheaper.

Additionally I think this doesn't need to return an `Option`; it can return `0` if there's no preceding newline.

sysint64 replied:
`\n` is 0x0a

cmyr commented on druid/src/text/editable_text.rs:
```diff
@@ -156,6 +162,32 @@ impl EditableText for String {
+            offset += grapheme.len();
+        }
+
+        Some(self.len())
```
Same as above, except that we can just return `self.len()` if we don't find a hit.

sysint64 commented:
@cmyr fixed.

cmyr commented on druid/src/text/editable_text.rs:
```diff
@@ -58,6 +58,12 @@ pub trait EditableText: Sized {
+    fn left_end_of_line(&self, offset: usize) -> usize;
+
+    /// Get the line ending offset from the given offset
+    fn right_end_of_line(&self, offset: usize) -> usize;
```
Ugh I'm sorry, I screwed up here, was thinking somehow of an API a level higher in the stack.

These should really be `preceding_line_break` and `next_line_break`. Whether or not they are on the left or the right will depend on  the writing direction.

cmyr commented on druid/src/text/movement.rs:
```diff
@@ -51,8 +55,11 @@ pub fn movement(m: Movement, s: Selection, text: &impl EditableText, modify: boo
-        Movement::LeftOfLine => 0,
-        Movement::RightOfLine => text.len(),
+        Movement::LeftOfLine => text.left_end_of_line(s.end),
+        Movement::RightOfLine => text.right_end_of_line(s.end),
```
These are also the wrong names, given the implementation. Ideally these *would* be the names and we would check writing direction on the paragraph, but we don't have that ability right now.

I think it's fine to leave this as is in this PR, and I'll open an issue describing the problem.

cmyr reviewed (commented):
Okay apologies, I was a little brain dead during my previous review. There's still a naming issue here, and a larger issue with the naming of the movement types themselves. I'd like to at least get the naming correct here, and we can address the larger issue separately. Sorry for the run-around!

sysint64 commented:
@cmyr it looks like there is an error on the master branch located in `druid/src/box_constraints.rs`

sysint64 commented:
this is what I get after rebase:
```
error[E0428]: the name `tests` is defined multiple times
   --> druid/src/box_constraints.rs:375:1
    |
276 | mod tests {
    | --------- previous definition of the module `tests` here
...
375 | mod tests {
    | ^^^^^^^^^ `tests` redefined here
    |
    = note: `tests` must be defined only once in the type namespace of this module
```

sysint64 commented:
fixed here: #1187

cmyr reviewed (approved):
Thanks!
